feat: async init indexdb

siuming-qiu opened the pull request:
I have encountered a performance issue in my project where some users experience a significant delay during the initialization of IndexDB, causing the rendering process to be blocked. To address this problem, I have implemented a solution to optimize the performance.

My approach involves introducing a variable to control the operation of IndexDB. When the createCacheProvider function is called, it will return a map object, and the operation of IndexDB will be locked. This means that any subsequent operations on IndexDB will be prevented until the initialization process is completed, and the lock is released.

By implementing this solution, I aim to improve the user experience by ensuring that the rendering process is not blocked by lengthy IndexDB initialization.

Commit:
feat: async init indexdb

piotr-cz commented on src/cache-provider.ts:
```diff
@@ -26,7 +26,9 @@ export default async function createCacheProvider<Data = any, Error = any>({
   // Initialize database
   let db: IDBPDatabase
 
-  try {
```
Please don't remove the `try...catch` block, it has been added here: #9 

Eventually it may be rewritten to this:
```ts
initDb().catch(error => 
  onError(error)
)
```

You should be able to reproduce exceptions by testing in Firefox Private mode, as described here: #5

piotr-cz commented:
Few questions:
- Do you have information on how is performance (database initialization and population) related to database size?
- What happens to cache writes that happen during database initialization?
- Should there be added a function callback parameter `onInit`?

Commit:
feat: add onInit and onError

siuming-qiu commented:
> Few questions:
> 
> * Do you have information on how is performance (database initialization and population) related to database size?
> * What happens to cache writes that happen during database initialization?
> * Should there be added a function callback parameter `onInit`?

- In my project, I have added some logging code as shown below. Occasionally, some users experience up to a 10-second delay during indexDB initialization, and I'm not sure why.
`const cacheProvider = useCacheProvider({
    dbName: DB_NAME,
    storeName: STORE_NAME,
    storageHandler: gcStorageHandler,
    version: DB_VERSION,
  });
if (!cacheProvider) {
    initDBTime = new Date().getTime();
    return null;
  }
  if (initDBTime) {
    const cost = new Date().getTime() - initDBTime;
    slardar.sendLog?.({
    extra: {
        cost,
      },
    });
    initDBTime = undefined;
  }`

- During database initialization, cache data is still written to the map but not stored in the database.
- Yes, that's a good idea, and I have added it.

piotr-cz commented:
I think this is not the proper solution for this problem.

- this doesn't address the problem of slow initialization, just hides it;
  cached data won't be accessible when app starts and this makes this package useless
- this adds a breaking change
- during initialization new data won't be stored in persistent cache

I'm using this package in apps to cache only few requests and database initialization and hydration takes less than a second.

I think you might be caching hundreds of requests and this is why it takes so much time to initialize.

Please take a look at implementing simple garbage collector: https://github.com/piotr-cz/swr-idb-cache?tab=readme-ov-file#implement-garbage-collector

piotr-cz commented:
I did simple performance check with 20× CPU slowdown in DevTools:
- Empty database: `useCacheProvider` takes ~700ms
- Database with large 9 GraphQL entries: `useCacheProvider` takes ~1000ms

I still think that garbage collector is the way to address this issue
